data-prepper-plugins/saas-source-plugins/jira-source/src/main/java/org/opensearch/dataprepper/plugins/source/jira/JiraService.java

```diff
@@ -0,0 +1,360 @@
+    private final Counter searchResultsFoundCounter;
+    private final Timer ticketFetchLatencyTimer;
+    private final Timer searchCallLatencyTimer;
+    private final PluginMetrics jiraPluginMetrics = PluginMetrics.fromNames("jiraService", "aws");
+
+
+    public JiraService(RestTemplate restTemplate,
+                       JiraSourceConfig jiraSourceConfig,
+                       JiraAuthConfig authConfig) {
+        this.restTemplate = restTemplate;
+        this.jiraSourceConfig = jiraSourceConfig;
+
+        issuesRequestedCounter = jiraPluginMetrics.counter(ISSUES_REQUESTED);
+        ticketFetchLatencyTimer = jiraPluginMetrics.timer(TICKET_FETCH_LATENCY_TIMER);
+        searchResultsFoundCounter = jiraPluginMetrics.counter(SEARCH_RESULTS_FOUND);
+        searchCallLatencyTimer = jiraPluginMetrics.timer(SEARCH_CALL_LATENCY_TIMER);
+        this.authConfig = authConfig;
+
+    }
+
+    /**
+     * Get jira entities.
+     *
+     * @param configuration       the configuration.
+     * @param timestamp           timestamp.
+     * @param itemInfoQueue       Item info queue.
+     * @param futureList          Future list.
+     * @param crawlerTaskExecutor Executor service.
+     */
+    public void getJiraEntities(JiraSourceConfig configuration, Instant timestamp,
+                                Queue<ItemInfo> itemInfoQueue, List<Future<Boolean>> futureList,
+                                ExecutorService crawlerTaskExecutor) {
+        log.info("Started to fetch entities");
+        jiraProjectCache.clear();
+        searchForNewTicketsAndAddToQueue(configuration, timestamp, itemInfoQueue, futureList, crawlerTaskExecutor);
+        log.trace("Creating item information and adding in queue");
+        jiraProjectCache.keySet().forEach(key -> {
+            Map<String, Object> metadata = new HashMap<>();
+            metadata.put(CONTENT_TYPE, JiraContentType.PROJECT.getType());
+            ItemInfo itemInfo = createItemInfo(_PROJECT + key, metadata);
+            itemInfoQueue.add(itemInfo);
+        });
+
+    }
+
+    /**
+     * Method for building Issue Item Info.
+     *
+     * @param configuration Input Parameter
+     * @param timestamp     Input Parameter
+     */
+    private void searchForNewTicketsAndAddToQueue(JiraSourceConfig configuration, Instant timestamp,
+                                                  Queue<ItemInfo> itemInfoQueue, List<Future<Boolean>> futureList,
+                                                  ExecutorService crawlerTaskExecutor) {
+        log.trace("Looking for Add/Modified tickets with a Search API call");
+        StringBuilder jql = createIssueFilterCriteria(configuration, timestamp);
+        int total;
+        int startAt = 0;
+        try {
+            do {
+                SearchResults searchIssues = getAllIssues(jql, startAt, configuration);
+                List<IssueBean> issueList = new ArrayList<>(searchIssues.getIssues());
+                total = searchIssues.getTotal();
```

Comment:
Shouldn't this be outside the loop?

Reply:
Initially, I thought of the same. In this case, it has to be inside the loop. In a paginated Jira response, each page contains the total number of issues that we are paginating. Also, we don't know this number before the first page.